File: rgctl-tests/ecommerce-java/src/main/java/com/example/ecommerce/service/ProductService.java
package com.example.ecommerce.service;

import com.example.ecommerce.dto.ProductDto;
import com.example.ecommerce.entity.Category;
import com.example.ecommerce.entity.Inventory;
import com.example.ecommerce.entity.Product;
import com.example.ecommerce.exception.ResourceNotFoundException;
import com.example.ecommerce.repository.ProductRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ProductService {

    private final ProductRepository productRepository;
    private final CategoryService categoryService;

    public ProductService(ProductRepository productRepository, CategoryService categoryService) {
        this.productRepository = productRepository;
        this.categoryService = categoryService;
    }

    @Transactional(readOnly = true)
    public List<ProductDto> findAll() {
        return productRepository.findAll().stream().map(this::toDto).toList();
    }

    @Transactional(readOnly = true)
    public List<ProductDto> findByCategory(Long categoryId) {
        return productRepository.findByCategoryId(categoryId).stream().map(this::toDto).toList();
    }

    @Transactional(readOnly = true)
    public ProductDto findById(Long id) {
        return toDto(getProduct(id));
    }

    @Transactional
    public ProductDto create(ProductDto dto) {
        Category category = categoryService.getCategory(dto.categoryId());
        Product product = new Product();
        product.setName(dto.name());
        product.setDescription(dto.description());
        product.setPrice(dto.price());
        product.setCategory(category);
        Inventory inventory = new Inventory();
        inventory.setProduct(product);
        inventory.setQuantity(dto.stockQuantity() != null ? dto.stockQuantity() : 0);
        product.setInventory(inventory);
        return toDto(productRepository.save(product));
    }

    public Product getProduct(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found: " + id));
    }

    private ProductDto toDto(Product product) {
        Integer stock = product.getInventory() != null ? product.getInventory().getQuantity() : null;
        return new ProductDto(
                product.getId(),
                product.getName(),
                product.getDescription(),
                product.getPrice(),
                product.getCategory().getId(),
                product.getCategory().getName(),
                stock
        );
    }
}

File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Services/ProductService.cs
using Ecommerce.Dto;
using Ecommerce.Models;
using Ecommerce.Exceptions;
using Ecommerce.Repositories;

namespace Ecommerce.Services;

public class ProductService
{
    private readonly ProductRepository _productRepository;
    private readonly CategoryService _categoryService;

    public ProductService(ProductRepository productRepository, CategoryService categoryService)
    {
        _productRepository = productRepository;
        _categoryService = categoryService;
    }

    public async Task<List<ProductDto>> FindAllAsync()
    {
        var products = await _productRepository.FindAllAsync();
        return products.Select(ToDto).ToList();
    }

    public async Task<List<ProductDto>> FindByCategoryAsync(long categoryId)
    {
        var products = await _productRepository.FindByCategoryIdAsync(categoryId);
        return products.Select(ToDto).ToList();
    }

    public async Task<ProductDto> FindByIdAsync(long id)
    {
        return ToDto(await GetProductAsync(id));
    }

    public async Task<ProductDto> CreateAsync(ProductDto dto)
    {
        _ = await _categoryService.GetCategoryAsync(dto.CategoryId);

        var product = new Product
        {
            Name = dto.Name,
            Description = dto.Description,
            Price = dto.Price,
            CategoryId = dto.CategoryId,
            Inventory = new Inventory
            {
                Quantity = dto.StockQuantity ?? 0
            }
        };

        product = await _productRepository.SaveAsync(product);
        return ToDto(await GetProductAsync(product.Id));
    }

    public async Task<Product> GetProductAsync(long id)
    {
        return await _productRepository.FindByIdAsync(id)
            ?? throw new ResourceNotFoundException($"Product not found: {id}");
    }

    private static ProductDto ToDto(Product product)
    {
        var stock = product.Inventory?.Quantity;
        return new ProductDto(
            product.Id,
            product.Name,
            product.Description,
            product.Price,
            product.CategoryId,
            product.Category.Name,
            stock);
    }
}
